### PR TITLE
Jenkins client for job enforcement

### PR DESCRIPTION
**Pull request by cznewt**

### CODE REVIEW

File: jenkins/client.sls [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{% from "jenkins/map.jinja" import client with context %}
+{%- if client.enabled %}
+
+jenkins_client_install:
+  pkg.installed:
+  - names: {{ client.pkgs }}
+
+jenkins_client_dirs:
+  file.directory:
+  - name: /srv/salt/env/dev/_jenkins_jobs
```
Comment (fpytloun) [inline jenkins/client.sls]: This is sooo ugly location, also jenkins.client doesn't need to run on salt master to use this path root dedicated to setup salt master env.
Maybe use `/var/tmp/salt_jenkins_jobs` or don't even use that temp file.

File: jenkins/files/_jenkins.conf [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{%- from "jenkins/map.jinja" import client with context %}
+jenkins:
+  url: "{{ client.master.get('proto', 'http') }}://{{ client.master.host }}:{{ client.master.port }}"
+  user: {{ client.master.username }}
```
Comment (fpytloun) [inline jenkins/files/_jenkins.conf:5]: Can be anonymous, make it optional.

File: jenkins/files/jobs/workflow.xml [hunk 3]
```diff
@@ -0,0 +1,43 @@
+{%- from "jenkins/map.jinja" import client with context %}
+{%- set job = salt['pillar.get']('jenkins:client:job:'+job_name) -%}
+<?xml version='1.0' encoding='UTF-8'?>
+<flow-definition plugin="workflow-job@2.5">
+  <description>{{ job.get('description', 'Salt generated, do not edit.') }}</description>
```
Comment (fpytloun) [inline jenkins/files/jobs/workflow.xml]: What about this? :)

> This job was generated using Salt jenkins formula. Any manual changes may be overwritten

File: jenkins/files/jobs/workflow.xml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+          {%- if param.type == "choice" %}
+          <choices class="java.util.Arrays$ArrayList">
+            <a class="string-array">
+              {%- for choice in param.choices %}
+              <string>{{ choice|e }}</string>
+              {%- endfor %}
+            </a>
+          </choices>
+          {%- endif %}
+          <defaultValue>{{ param.default|e }}</defaultValue>
```
Comment (fpytloun) [inline jenkins/files/jobs/workflow.xml:42]: Default value should be optional.

File: README.rst [hunk 5]
```diff
@@ -110,6 +120,72 @@ Jenkins slave
...
+              node {
+                 stage 'Stage 1'
+                 echo 'Hello World 1'
+                 stage 'Stage 2'
+                 echo 'Hello World 2'
+              }
+          test_workflow_jenkins_input:
+            type: workflow
+            display_name: Test jenkins workflow inputs
+            script: |
```
Comment (fpytloun) [inline README.rst]: There should be support for deploying workflow job which will use SCM repository to get scripts. This is good for smaller ones but you cannot ship more complex with includable modules, etc.

Comment (cznewt) [reply README.rst]: yes, this apporach doesnt' work, the interpolation collides :disappointed:

File: jenkins/client.sls (hunk 1 above)
Comment (cznewt) [reply jenkins/client.sls]: Needs to be in salt file root, soooo only thing ugly here is it's not being parametrised :wink: 
I'll into it, and `/var/cache/salt` might be even better

File: README.rst [hunk 6]
```diff
@@ -110,6 +120,93 @@ Jenkins slave
...
+              }
+
+
+GIT controlled groovy script samples
+
+.. code-block:: yaml
+
+    jenkins:
+      client:
+        source:
```
Comment (fpytloun) [inline README.rst:209]: There can be separate repositories per-workflow. Eg. docker build workflow can be from different git repository than another one.

Comment (michaelkuty) [reply README.rst:209]: KISS why you want to have repo / ,,script" ?

**Pull request merged**